### ISSUE TITLE
Add option "passwordIgnore" to TOTP authenticator

### ISSUE THREAD

**Issue by goodomens42** [opnsense/core#4820]
- [ x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [x ] I have searched the existing issues and I am convinced that mine is new.

OPNsense provides TOTP support for different services, for example OpenVPN user dialin.
In order to use TOTP you have to specify the TOTP token followed by the user password instead of the user password only.
This might lead to tedious typing which maybe can not be completed correctly within TOTP's 30s window and therefore little acceptance of TOTP by the affected users.

For some use cases it might be sufficient to supply only the TOTP token as authentification.
I have created a small patch that adds an extra option "passwordIgnore" to the TOTP authenticator.
When setting this option to true, the authenticator will be happy with the TOTP token only and ignore the user password.
The default value for this extra option is "false" in order not to break existing behaviour.

We have been using this patch for OpenVPN dialin in production use for a long time now.

[opnsense_totp.patch.txt](https://github.com/opnsense/core/files/6137254/opnsense_totp.patch.txt)

**Comment by AdSchellevis** [comment]
It's pretty common to use totp as an extra factor for authentication, you could always build a "token only" authenticator for your own use (inherit the totp part and ignore the password, more information about the framework is available here https://docs.opnsense.org/development/components/authentication.html), we're just not going to include this in our product for security reasons.

**Comment by fichtner** [comment]
I have to agree with Ad here. Main problems are that when the phone with the TOTP secret has been snatched or when the attacker can control the NTP server of your firewall you are completely vulnerable.

> This might lead to tedious typing which maybe can not be completed correctly within TOTP's 30s window and therefore little acceptance of TOTP by the affected users.

For long passwords it's best to allow the password first since you can finish typing and then focus on the OTP.


Cheers,
Franco

**Comment by AdSchellevis** [comment]
in case of OpenVPN I think some clients support saving and splitting the password and token as well (https://github.com/opnsense/core/issues/3290)

**Comment by goodomens42** [comment]
Just to make myself clear:
I'm using that as setable switch in the TOTP authenticators configuration

![Screenshot 2021-03-14 204510](https://user-images.githubusercontent.com/60742141/111082040-7e49dd00-8506-11eb-830f-82113f9d7d04.jpg)

Everybody who is concerned about security needn't tick it and "ignoring the password" ist not the default.
In our production setup we are in fact using two LDAP-TOTP authenticators, one with "passwordIgnore" set, the other with default settings.

@AdSchellevis 
Building a seperate "Token Authenticator" is an other option for sure.
The nice thing about my "passwordIgnore" option is, that it works with the TOTP authenticators in OPNSense (Local Database + LDAP) out of the box.

**Comment by AdSchellevis** [comment]
I know what it looks like, we're just not going to add this type of functionality. Adding a separate authenticated makes it less obscure, but won't make it in our standard product either for security reasons (the latter is easier to manage on your end, so upgrades don't overwrite your local changes).

**Comment by NlightN22** [comment]
@goodomens42 Thank you very match! You make my life easy with your patch! This is not safely, i understand that, but i have another layer of safety, and for ovpn user i need only OTP pass.

**Comment by Gaitonde007** [comment]
> * [ x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
> * [x ] I have searched the existing issues and I am convinced that mine is new.
> 
> OPNsense provides TOTP support for different services, for example OpenVPN user dialin. In order to use TOTP you have to specify the TOTP token followed by the user password instead of the user password only. This might lead to tedious typing which maybe can not be completed correctly within TOTP's 30s window and therefore little acceptance of TOTP by the affected users.
> 
> For some use cases it might be sufficient to supply only the TOTP token as authentification. I have created a small patch that adds an extra option "passwordIgnore" to the TOTP authenticator. When setting this option to true, the authenticator will be happy with the TOTP token only and ignore the user password. The default value for this extra option is "false" in order not to break existing behaviour.
> 
> We have been using this patch for OpenVPN dialin in production use for a long time now.
> 
> [opnsense_totp.patch.txt](https://github.com/opnsense/core/files/6137254/opnsense_totp.patch.txt)

Hey, Thanks for such wonderful path. However after applying the patch authentication is still failing. I have enabled Ignore Password from GUI and also restarted PHP.

Is there anything else I have to do? Sorry because I'm network guy facing thing issue don't have idea about coding.

Regards,

**Comment by Gaitonde007** [comment]
> * [ x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
> * [x ] I have searched the existing issues and I am convinced that mine is new.
> 
> OPNsense provides TOTP support for different services, for example OpenVPN user dialin. In order to use TOTP you have to specify the TOTP token followed by the user password instead of the user password only. This might lead to tedious typing which maybe can not be completed correctly within TOTP's 30s window and therefore little acceptance of TOTP by the affected users.
> 
> For some use cases it might be sufficient to supply only the TOTP token as authentification. I have created a small patch that adds an extra option "passwordIgnore" to the TOTP authenticator. When setting this option to true, the authenticator will be happy with the TOTP token only and ignore the user password. The default value for this extra option is "false" in order not to break existing behaviour.
> 
> We have been using this patch for OpenVPN dialin in production use for a long time now.
> 
> [opnsense_totp.patch.txt](https://github.com/opnsense/core/files/6137254/opnsense_totp.patch.txt)

Hey, Thanks for such wonderful patch. However after applying the patch authentication is still failing. I have enabled Ignore Password from GUI and also restarted PHP.

Is there anything else I have to do? Sorry because I'm network guy facing this issue don't have idea about coding.

Regards
